pkg/controller/direct/compute/networkedgesecurityservice_controller.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &NetworkEdgeSecurityServiceAdapter{
+		gcpClient: networkEdgeSecurityServicesClient,
+		id:        id,
+		desired:   obj,
+		reader:    reader,
+	}, nil
+}
+
+func (m *modelNetworkEdgeSecurityService) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type NetworkEdgeSecurityServiceAdapter struct {
+	gcpClient *compute.NetworkEdgeSecurityServicesClient
+	id        *krm.NetworkEdgeSecurityServiceIdentity
+	desired   *krm.ComputeNetworkEdgeSecurityService
+	actual    *computepb.NetworkEdgeSecurityService
+	reader    client.Reader
+}
+
+var _ directbase.Adapter = &NetworkEdgeSecurityServiceAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *NetworkEdgeSecurityServiceAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting ComputeNetworkEdgeSecurityService", "name", a.id)
+
+	req := &computepb.GetNetworkEdgeSecurityServiceRequest{
+		Project:                    a.id.Parent().ProjectID,
+		Region:                     a.id.Parent().Location,
+		NetworkEdgeSecurityService: a.id.ID(),
+	}
+	actual, err := a.gcpClient.Get(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeNetworkEdgeSecurityService %q: %w", a.id, err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *NetworkEdgeSecurityServiceAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating ComputeNetworkEdgeSecurityService", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := ComputeNetworkEdgeSecurityServiceSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Name = direct.LazyPtr(a.id.ID())
+
+	req := &computepb.InsertNetworkEdgeSecurityServiceRequest{
+		Project:                            a.id.Parent().ProjectID,
+		Region:                             a.id.Parent().Location,
+		NetworkEdgeSecurityServiceResource: resource,
+	}
+	op, err := a.gcpClient.Insert(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating ComputeNetworkEdgeSecurityService %s: %w", a.id, err)
+	}
+
+	err = op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("compute ComputeNetworkEdgeSecurityService %s waiting creation: %w", a.id.String(), err)
+	}
+
+	log.V(2).Info("successfully created ComputeNetworkEdgeSecurityService", "name", a.id)
+
+	// Get the created resource
+	created := &computepb.NetworkEdgeSecurityService{}
```

Comment:
We are fetching the resource in Find, Create and Update. Maybe put the shared logic into a separate function?